tests/expressions/tests.py

```diff
@@ -391,6 +401,216 @@ def test_ticket_18375_chained_filters(self):
+        Company.objects.create(name="5020 Ltd", num_employees=50, num_chairs=20, ceo=ceo)
+        Company.objects.create(name="5040 Ltd", num_employees=50, num_chairs=40, ceo=ceo)
+        Company.objects.create(name="5050 Ltd", num_employees=50, num_chairs=50, ceo=ceo)
+        Company.objects.create(name="5060 Ltd", num_employees=50, num_chairs=60, ceo=ceo)
+        Company.objects.create(name="99300 Ltd", num_employees=99, num_chairs=300, ceo=ceo)
+
+    def setUp(self):
+        self.company_query = Company.objects.values(
+            "name", "num_employees", "num_chairs"
+        ).order_by(
+            "name", "num_employees", "num_chairs"
+        )
+
+    def test_in_lookup_allows_F_expressions_and_expressions_for_integers(self):
+        # Test that in lookups can use F() expressions for integers
+        queryset = Company.objects.filter(
+            num_employees__in=([F('num_chairs') - 10])
+        )
+        self.assertQuerysetEqual(
+            queryset, [
+                '<Company: 5060 Ltd>'
+            ],
+            ordered=False
+        )
+
+        queryset = Company.objects.filter(
+            num_employees__in=([F('num_chairs') - 10, F('num_chairs') + 10])
+        )
+        self.assertQuerysetEqual(
+            queryset,
+            [
+                '<Company: 5040 Ltd>',
+                '<Company: 5060 Ltd>'
+            ],
+            ordered=False
+        )
+
+        queryset = Company.objects.filter(
+            num_employees__in=([F('num_chairs') - 10, F('num_chairs'), F('num_chairs') + 10])
+        )
+        self.assertQuerysetEqual(
+            queryset,
+            [
+                '<Company: 5040 Ltd>',
+                '<Company: 5050 Ltd>',
+                '<Company: 5060 Ltd>'
+            ],
+            ordered=False
+        )
+
+    def test_allowing_expressions_in_lookups_does_not_cause_incorrect_join_choice(self):
+        start_time = datetime.time(12, 0)
+        end_time = datetime.time(14, 0)
+        actual = datetime.time(13, 0)
+        t1 = Time.objects.create(time=start_time)
+        t2 = Time.objects.create(time=end_time)
+
+        SimulationRun.objects.create(start=t1, end=t2, midpoint=actual)
+        SimulationRun.objects.create(start=t1, end=None, midpoint=actual)
+        SimulationRun.objects.create(start=None, end=t2, midpoint=actual)
+        SimulationRun.objects.create(start=None, end=None, midpoint=actual)
+
+        queryset = SimulationRun.objects.filter(midpoint__range=[F('start__time'), F('end__time')])
+        self.assertQuerysetEqual(
+            queryset,
+            [
+                '<SimulationRun: 13:00:00 (12:00:00 to 14:00:00)>',
+            ],
+            ordered=False
+        )
+        for alias in queryset.query.alias_map.values():
+            if isinstance(alias, Join):
+                self.assertEqual(alias.join_type, constants.INNER)
```

Comment:
Based on the conversation around https://github.com/django/django/pull/6216#discussion_r54401771 shouldn't these joins be LEFT since the start and end fields are nullable? The result is correct in this case but it wouldn't be if a query relied on the join being nullable. I'm struggling to come up with a query to exercise the problem. @akaariai you have anything for this?